syntax = "proto3";

package proto;

import "vector_math.proto";

enum TextSizeEnum {
    TEXT_SIZE_SMALL = 0;    // 18 pixels
    TEXT_SIZE_MEDIUM = 1;   // 24 pixels
    TEXT_SIZE_LARGE = 2;    // 32 pixels
    TEXT_SIZE_TITLE = 3;    // 96 pixels
}

// FontSize
// Next: 3
message FontSize {
    TextSizeEnum text_size_enum = 1;
    double size = 2;
}

// ClientParameter
// Next: 7
message ClientParameter {
    // Default server name.
    string server_name = 1;
    // Default font.
    string font_file = 2;
    // Default font sizes.
    repeated FontSize font_sizes = 3;
    // Overlay of the title.
    PageDescription overlay_title = 4;
    // This is the overlay on state change.
    PageDescription overlay_state = 5;
    // This is the overlay when you play.
    PageDescription overlay_play = 6;
}

// PageElementText
// Next: 5
message PageElementText {
    string text = 1;
    Vector4 color = 2;
    Vector2 position = 3;
    TextSizeEnum text_size_enum = 4;
}

// PageElementImage
// Next: 4
message PageElementImage {
    string image = 1;
    Vector2 position = 2;
    Vector2 size = 3;
}

// PageElementLine
// Next: 5
message PageElementLine {
    Vector2 start = 1;
    Vector2 end = 2;
    Vector4 color = 3;
    double thickness = 4;
}

// PageElementRectFilled
// Next: 5
message PageElementRectFilled {
    Vector2 position = 1;
    Vector2 size = 2;
    Vector4 color = 3;
    double rounding = 4;
}

// PageElement
// Next: 5
message PageElement {
    oneof PageElementOneof {
        PageElementText text = 1;
        PageElementImage image = 2;
        PageElementLine line = 3;
        PageElementRectFilled rect_filled = 4;
    }
}

// PageDescription
// Next: 4
message PageDescription {
    string name = 1;
    Vector2 size = 2;
    repeated PageElement page_elements = 3;
}
